File: gravitee-apim-console-webui/src/services/role.service.ts
/*
 * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as _ from 'lodash';

class RoleService {
  private permissionsByScope: any;

  constructor(private $http, private Constants, private $q) {
    'ngInject';
  }

  listRights() {
    return ['c', 'r', 'u', 'd'];
  }

  listScopes() {
    return this.fetchScopes().then((permissionsByScope) => {
      return permissionsByScope;
    });
  }

  listPermissionsByScope(scope: string) {
    return this.fetchScopes().then((permissionsByScope) => {
      return permissionsByScope[scope];
    });
  }

  isUserRoleManagement(scope: string) {
    return 'ORGANIZATION' === scope;
  }

  get(roleScope, roleName) {
    return this.$http.get(`${this.Constants.org.baseURL}/configuration/rolescopes/` + roleScope + '/roles/' + roleName).then((response) => {
      const role = response.data;
      role.scope = _.toUpper(role.scope);
      return role;
    });
  }

  list(scope: string) {
    return this.$http.get(`${this.Constants.org.baseURL}/configuration/rolescopes/` + scope + '/roles').then((response) => {
      return _.map(response.data, (role: any) => {
        role.scope = _.toUpper(role.scope);
        return role;
      });
    });
  }

  create(role) {
    return this.$http.post(`${this.Constants.org.baseURL}/configuration/rolescopes/` + role.scope + '/roles', role).then((response) => {
      const role = response.data;
      role.scope = _.toUpper(role.scope);
      return role;
    });
  }

  update(role) {
    return this.$http
      .put(`${this.Constants.org.baseURL}/configuration/rolescopes/` + role.scope + '/roles/' + role.name, role)
      .then((response) => {
        const role = response.data;
        role.scope = _.toUpper(role.scope);
        return role;
      });
  }

  delete(role) {
    return this.$http.delete(`${this.Constants.org.baseURL}/configuration/rolescopes/` + role.scope + '/roles/' + role.name);
  }

  listUsers(roleScope, roleName) {
    return this.$http
      .get(`${this.Constants.org.baseURL}/configuration/rolescopes/` + roleScope + '/roles/' + roleName + '/users')
      .then((response) => response.data);
  }

  deleteUser(role, username) {
    return this.$http.delete(
      `${this.Constants.org.baseURL}/configuration/rolescopes/` + role.scope + '/roles/' + role.name + '/users/' + username,
    );
  }

  addRole(roleScope, roleName, user) {
    return this.$http.post(`${this.Constants.org.baseURL}/configuration/rolescopes/${roleScope}/roles/${roleName}/users`, user);
  }

  private fetchScopes() {
    if (this.permissionsByScope) {
      return this.$q.resolve(this.permissionsByScope);
    } else {
      return this.$http.get(`${this.Constants.org.baseURL}/configuration/rolescopes/`).then((response) => {
        this.permissionsByScope = response.data;
        return this.permissionsByScope;
      });
    }
  }
}

export default RoleService;

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/RoleService.java
/*
 * Copyright © 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.gravitee.rest.api.service;

import io.gravitee.rest.api.model.MembershipReferenceType;
import io.gravitee.rest.api.model.NewRoleEntity;
import io.gravitee.rest.api.model.RoleEntity;
import io.gravitee.rest.api.model.UpdateRoleEntity;
import io.gravitee.rest.api.model.permissions.Permission;
import io.gravitee.rest.api.model.permissions.RolePermissionAction;
import io.gravitee.rest.api.model.permissions.RoleScope;
import io.gravitee.rest.api.model.permissions.SystemRole;
import io.gravitee.rest.api.service.common.ExecutionContext;
import java.util.*;

/**
 * @author Nicolas GERAUD (nicolas.geraud at graviteesource.com)
 * @author Florent CHAMFROY (forent.chamfroy at graviteesource.com)
 * @author GraviteeSource Team
 */
public interface RoleService {
    RoleEntity create(ExecutionContext executionContext, final NewRoleEntity roleEntity);
    void createOrUpdateSystemRole(
        ExecutionContext executionContext,
        SystemRole roleName,
        RoleScope roleScope,
        Permission[] permissions,
        String organizationId
    );
    void createOrUpdateSystemRoles(ExecutionContext executionContext, String organizationId);
    void delete(ExecutionContext executionContext, String roleId);
    List<RoleEntity> findAllByOrganization(String organizationId);
    RoleEntity findById(String roleId);
    Set<RoleEntity> findAllById(Set<String> roleIds);
    Optional<RoleEntity> findByIdAndOrganizationId(String roleId, String organizationId);
    List<RoleEntity> findByScope(RoleScope scope, String organizationId);
    Optional<RoleEntity> findByScopeAndName(RoleScope scope, String name, String organizationId);
    List<RoleEntity> findDefaultRoleByScopes(String organizationId, RoleScope... scopes);
    boolean hasPermission(Map<String, char[]> userPermissions, Permission permission, RolePermissionAction[] acls);
    void initialize(ExecutionContext executionContext, String organizationId);
    RoleEntity update(ExecutionContext executionContext, UpdateRoleEntity role);
    RoleScope findScopeByMembershipReferenceType(MembershipReferenceType type);
    RoleEntity findPrimaryOwnerRoleByOrganization(String organizationId, RoleScope roleScope);
}
